File: runtimejs/java/util/HashSet.js
//reference// java/util/HashMap
//load// java/util/Set
//load// java/util/AbstractCollection
//complete// java/lang/Object
var java_util_HashSet = function() 
{   this._map = null;
};

_defclass(java_util_HashSet, java_util_AbstractCollection, [java_util_Set], 
{   _0: function() 
    {   this._map = new java_util_HashMap()._0();
        return this;
    },
        
    _1Ljava_util_Collection$: function(collection) 
    {   this._0();
        this.addAll_1(collection);
        return this;
    },
        
    add_1: function(e) 
    {   return this._map.put_2(e, java_util_HashSet_PRESENT)===null;
    },           
        
    // OPTIMIZATION
    clear_0: function() 
    {   this._map.clear_0();
    },
        
    // OPTIMIZATION
    contains_1: function(e) 
    {   return this._map.containsKey_1(e);
    },

    // containsAll        implemented by AbstractCollection
                    
    equals_1: function(o) 
    {   if (o===null || !_isinterface(o,java_util_Set)) return false;            
        var s = o;
        if (this.size_0() !== s.size_0()) return false;
        for (var it=s.iterator_0(); it.hasNext_0(); ) 
        {   if (! this.contains_1(it.next_0())) return false;
        }            
        return true; 
    },
                
    hashCode_0: function() 
    {   return this._map.keySet_0().hashCode_0();
    },
        
    // isEmpty       implemented by AbstractCollection
        
    iterator_0: function() 
    {   return this._map.keySet_0().iterator_0();
    },

    // OPTIMIZATION
    remove_1Ljava_lang_Object$: function(key) 
    {   return this._map.remove_1(key)!==null;
    },
        
    // OPTIMIZATION
    removeAll_1: function(collection) 
    {   var i = collection.iterator_0();
        var didremove = false;
        while (i.hasNext_0()) 
        {   if (this.remove_1Ljava_lang_Object$(i.next_0())) 
            {   didremove = true;
            }
        }        
        return didremove;            
    },
        
    size_0: function() 
    {   return this._map.size_0();
    }
        
    // toArray   implemented by AbstractCollection 
    // toString  implemented by AbstractCollection
});

var java_util_HashSet_PRESENT = new java_lang_Object()._0();  

File: runtimecs/java/util/HashSet.cs
using java.lang;

namespace java.util 
{ 
    public class HashSet : AbstractCollection, Set
    {
        private static readonly object PRESENT = new object();
        private readonly HashMap map;
            
        public HashSet() 
        {   
            map = new HashMap();
        }
            
        public HashSet(Collection c) : this() 
        {   
            addAll(c);
        }
            
        public override bool add(object e) 
        {   
            return map.put(e, PRESENT)==null;
        }           

        // OPTIMIZATION    
        public override void clear() 
        {   
            map.clear();
        }
            
        // OPTIMIZATION
        public override bool contains(object e) 
        {   
            return map.containsKey(e);
        }

        // containsAll        implemented by AbstractCollection
                        
        public override bool Equals(object o) 
        {   
            if (o==null || !(o is Set)) { return false; }
            Set s = (Set) o;
            if (size() != s.size()) { return false; }
            for (Iterator it=s.iterator(); it.hasNext(); ) 
            {   if (! contains(it.next())) { return false; }
            }            
            return true; 
        }
                    
        public override int GetHashCode() 
        {   
            return map.keySet().GetHashCode();
        }
            
        // isEmpty       implemented by AbstractCollection
            
        public override Iterator iterator() 
        {   
            return map.keySet().iterator();
        }
        

        // OPTIMIZATION    
        public override bool remove(object key) 
        {   
            return map.remove(key)!=null;
        }
        
        // OPTIMIZATION
        public override bool removeAll(Collection collection) 
        {   
            Iterator i = collection.iterator();
            bool didremove = false;
            while (i.hasNext()) 
            {   
                if (this.remove(i.next())) 
                {   didremove = true;
                }
            }        
            return didremove;            
        }        
            
        public override int size() 
        {   
            return map.size();
        }
           
        // toArray             implemented by AbstractCollection 
        // toArray(Object[])   implemented by AbstractCollection 
        // toString            implemented by AbstractCollection
    }
}
